esp/src/src-react/util/history.ts

```diff
@@ -0,0 +1,148 @@
+import UniversalRouter from "universal-router";
+import { parse, ParsedQuery, stringify } from "query-string";
+import { hashSum } from "@hpcc-js/util";
+
+let g_router: UniversalRouter;
+
+export function initialize(routes) {
+    g_router = new UniversalRouter(routes);
+    return g_router;
+}
+
+function parseHash(hash: string): HistoryLocation {
+    if (hash[0] !== "#") {
+        return {
+            pathname: "/",
+            search: "",
+            id: hashSum("#/")
+        };
+    }
+
+    const parts = hash.substring(1).split("?");
+    return {
+        pathname: parts[0],
+        search: parts.length > 1 ? `?${parts[1]}` : "",
+        id: hashSum(hash)
+    };
+}
+
+export function parseSearch<T = ParsedQuery<string | boolean | number>>(_: string): T {
+    if (_[0] !== "?") return {} as T;
+    return { ...parse(_.substring(1), { parseBooleans: true, parseNumbers: true }) } as unknown as T;
+}
+
+interface HistoryLocation {
+    pathname: string;
+    search: string;
+    id: string;
+}
+
+export type ListenerCallback<S extends object = object> = (location: HistoryLocation, action: string) => void;
+
+const globalHistory = globalThis.history;
+
+class History<S extends object = object> {
+
+    location: HistoryLocation = {
+        pathname: "/",
+        search: "",
+        id: hashSum("#/")
+    };
+    state: S = {} as S;
+
+    constructor() {
+        this.location = parseHash(document.location.hash);
+
+        window.addEventListener("hashchange", ev => {
+            console.log("hashchange: " + document.location);
```

Comment:
Will we have a global log helper? It would be nice to leave this type of thing in place but just enable/disable as needed.